### DISCUSSION TITLE
Netbox 2.11.3 and ubuntu 18.04 update to the last version.

### DISCUSSION THREAD

**Discussion by DocX39** [netbox-community/netbox#10113] [Q&A]
I am currently on an Ubuntu 18.04 server and Netbox 2.11.3.

**Dependencies:**
- Python 2.7.17
- Python 3.6.9
- postgresql 10.22
- Redis 4.0.9

Could you please guide me through the process of upgrading Netbox to version 3.x and Ubuntu 22.04?

Should I upgrade the operating system first or not?

Thanks in advance for your help.

**Comment by jeremystretch** [comment]
> Could you please guide me through the process of upgrading Netbox to version 3.x and Ubuntu 22.04?

It's possible to upgrade to the latest NetBox release (v3.3.0) directly from here simply following the normal [upgrade instructions](https://docs.netbox.dev/en/stable/installation/upgrading/).

> Should I upgrade the operating system first or not?

I would (actually I'd rebuild it entirely), but that's entirely your call.

**Reply by DocX39** [reply]
I set up a new ubuntu 22.04 server with Netbox 3.3.0.
Is it easy to migrate the PostgreSQL database from my production Netbox 2.11.3 to the new server with Netbox 3.3.0?

**Reply by candlerb** [reply]
Yes: https://docs.netbox.dev/en/stable/administration/replicating-netbox/

Restore the old database, then re-run the upgrade.sh script.  Also remember to copy any media items, scripts and reports.

**Reply by DocX39** [reply]
First of all, thank you everyone for your answers.

I have migrated the database from Netbox version 2.11.3 to version 3.3.0 which is on a new VM.

**After running the script :**

`sudo ./upgrade.sh`

**I get these logs :**

Running migrations:
  Applying tenancy.0002_tenant_ordering... OK
  Applying extras.0060_customlink_button_class... OK
  Applying extras.0061_extras_change_logging... OK
  Applying extras.0062_clear_secrets_changelog... OK
  Applying tenancy.0003_contacts... OK
  Applying tenancy.0004_extend_tag_support... OK
  Applying ipam.0047_prefix_depth_children... OK
  Applying ipam.0048_prefix_populate_depth_children...
Updating 743 prefixes...
 OK
  Applying ipam.0049_prefix_mark_utilized... OK
  Applying ipam.0050_iprange... OK
  Applying ipam.0051_extend_tag_support... OK
  Applying extras.0063_webhook_conditions... OK
  Applying extras.0064_configrevision... OK
  Applying ipam.0052_fhrpgroup... OK
  Applying ipam.0053_asn_model... OK
  Applying dcim.0132_cable_length... OK
  Applying dcim.0133_port_colors... OK
  Applying dcim.0134_interface_wwn_bridge... OK
  Applying dcim.0135_tenancy_extensions... OK
  Applying dcim.0136_device_airflow... OK
  Applying dcim.0137_relax_uniqueness_constraints... OK
  Applying dcim.0138_extend_tag_support... OK
  Applying dcim.0139_rename_cable_peer... OK
  Applying wireless.0001_wireless... OK
  Applying dcim.0140_wireless... OK
  Applying dcim.0141_asn_model... OK
  Applying dcim.0142_rename_128gfc_qsfp28... OK
  Applying dcim.0143_remove_primary_for_related_name... OK
  Applying dcim.0144_fix_cable_abs_length... OK
  Applying dcim.0145_site_remove_deprecated_fields...Traceback (most recent call last):
  File "/opt/netbox/netbox/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/core/management/base.py", line 414, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/core/management/base.py", line 460, in execute
    output = self.handle(*args, **options)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/core/management/base.py", line 98, in wrapped
    res = handle_func(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/core/management/commands/migrate.py", line 290, in handle
    post_migrate_state = executor.migrate(
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 131, in migrate
    state = self._migrate_all_forwards(
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 163, in _migrate_all_forwards
    state = self.apply_migration(
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 248, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/migrations/migration.py", line 131, in apply
    operation.database_forwards(
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/opt/netbox/netbox/dcim/migrations/0145_site_remove_deprecated_fields.py", line 24, in check_legacy_data
    raise DataError(
django.db.utils.DataError: Unable to proceed with deleting contact fields from Site model: Found 3 sites with legacy contact data. Please ensure all legacy site contact data has been migrated to contact objects before proceeding. Or, set the NETBOX_DELETE_LEGACY_DATA environment variable to bypass this safeguard and delete all legacy site contact data.



**The message on the home page of the web server:**
![Server Error](https://user-images.githubusercontent.com/49805674/186417221-58a24848-e9c5-4be8-b5d8-6424c5dec0db.png)

Do you see where the problem could come from?

**Reply by candlerb** [reply]
Yes, it's right there in the error message itself:

> Unable to proceed with deleting contact fields from Site model: Found 3 sites with legacy contact data. Please ensure all legacy site contact data has been migrated to contact objects before proceeding. Or, set the NETBOX_DELETE_LEGACY_DATA environment variable to bypass this safeguard and delete all legacy site contact data.

As explained in the [release notes for Netbox 3.2](https://github.com/netbox-community/netbox/releases/tag/v3.2.0), certain fields have been removed and are not automatically migrated to new models - specifically, the contact details on a site (there is now a separate 'contacts' model).

If you're happy to throw that data away, set NETBOX_DELETE_LEGACY_DATA=1 in your environment and re-run the migration.

If you don't want to throw that data away, then:
* migrate to 3.1
* move the contact data into the new Contacts model, either by hand or by using the script which was posted to this group (and set the old fields in the Site model to empty)
* migrate to 3.3

**Comment by mtinberg** [comment]
You need to update the OS first to get a newer Python 3.8 interpreter to build the new virtualenv.  Most important is backup the PostgreSQL database with pg_dump and any other static files, config, the whole Netbox install if you can, so you can always start over and try again if something goes wrong or you make a mistake during the upgrade.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: DocX39 ***@***.***>
Sent: Tuesday, August 23, 2022 9:38 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Netbox 2.11.3 and ubuntu 18.04 update to the last version. (Discussion #10113)


I am currently on an Ubuntu 18.04 server and Netbox 2.11.3.

Dependencies:

  *   Python 2.7.17
  *   Python 3.6.9
  *   postgresql 10.22
  *   Redis 4.0.9

Could you please guide me through the process of upgrading Netbox to version 3.x and Ubuntu 22.04?

Should I upgrade the operating system first or not?

Thanks in advance for your help.

**Reply by candlerb** [reply]
> You need to update the OS first to get a newer Python 3.8 interpreter to build the new virtualenv.

Actually, no you don't need to upgrade the OS.  But you do need to install python3.8, which is in the standard Ubuntu repositories for 18.04:

```
apt-get install python3.8 python3.8-venv
```

and you need to run the upgrade.sh slightly differently (every time):

```
PYTHON=python3.8 ./upgrade.sh
```

If you're running Netbox in a VM, then I'd say it's safer to make a fresh Netbox instance, import the pgdump from the original one, and upgrade that (as you can test this as much as you like while the old system remains running).  But an in-place upgrade from NB 2.11 to NB 3.3 *is* still possible on Ubuntu 18.04.

One other consideration is that some model fields were dropped between Netbox 3.1 and [3.2](https://github.com/netbox-community/netbox/releases/tag/v3.2.0) - in particular, the contact information on the Site model, since there's now a separate Contacts model. If that's a concern, then you should upgrade to 3.1.x first, then sort out moving the data into the right place, and then upgrade to 3.3.

As usual: the REST API does change between versions, so if you have any integrations which use them, or any custom scripts or reports, they may need changing.  That's another reason for running a separate 3.3 instance in a test environment, before you go live.

**Comment by mtinberg** [comment]
You can read through the upgrade instructions, but the main thing will be to load the database dump from the old Netbox 2.11 instance into the new Netbox 3.3 one, then run the upgrade script which will run database migrations and other maintenance tasks to make it compatible with Netbox 3.3.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
